benchmarks/asv_delegated.py

```diff
@@ -0,0 +1,350 @@
+        parentage = {k: v for k, v in parentage.items() if v is not None}
+        if len(parentage) == 0:
+            message = f"No env prep script available for commit: {commit_hash} ."
+            raise KeyError(message)
+        else:
+            parentage = dict(sorted(parentage.items(), key=lambda item: item[1]))
+            commit = next(iter(parentage))
+            content = self[commit]
+            return content
+
+
+class Delegated(Environment):
+    """Manage a benchmark environment using custom user scripts, run at each commit.
+
+    Ignores user input variations - ``matrix`` / ``pythons`` /
+    ``exclude``, since environment is being managed outside ASV.
+
+    A vanilla :class:`asv.environment.Environment` is created for containing
+    the expected ASV configuration files and checked-out project. The actual
+    'functional' environment is created/updated using the command(s) specified
+    in the config ``delegated_env_commands``, then the location is recorded via
+    a symlink within the ASV environment. The symlink is used as the
+    environment path used for any executable calls (e.g.
+    ``python my_script.py``).
+
+    """
+
+    tool_name = "delegated"
+    """Required by ASV as a unique identifier of the environment type."""
+
+    DELEGATED_LINK_NAME = "delegated_env"
+    """The name of the symlink to the delegated environment."""
+
+    COMMIT_ENVS_VAR = "ASV_COMMIT_ENVS"
+    """Env var that instructs a dedicated environment be created per commit."""
+
+    def __init__(self, conf, python, requirements, tagged_env_vars):
+        """Get a 'delegated' environment based on the given ASV config object.
+
+        Parameters
+        ----------
+        conf : dict
+            ASV configuration object.
+
+        python : str
+            Ignored - environment management is delegated. The value is always
+            ``DELEGATED``.
+
+        requirements : dict (str -> str)
+            Ignored - environment management is delegated. The value is always
+            an empty dict.
+
+        tagged_env_vars : dict (tag, key) -> value
+            Ignored - environment management is delegated. The value is always
+            an empty dict.
+
+        Raises
+        ------
+        EnvironmentUnavailable
+            The original environment or delegated environment cannot be created.
+
+        """
+        ignored = []
+        if python:
+            ignored.append(f"{python=}")
+        if requirements:
+            ignored.append(f"{requirements=}")
+        if tagged_env_vars:
+            ignored.append(f"{tagged_env_vars=}")
+        message = (
+            f"Ignoring ASV setting(s): {', '.join(ignored)}. Benchmark "
+            "environment management is delegated to third party script(s)."
+        )
+        log.warning(message)
+        self._python = "DELEGATED"
+        self._requirements = {}
+        self._tagged_env_vars = {}
+        super().__init__(
```

Comment:
Opportunity for some defensive typing on `self._path` before `super().__init__()`:

https://github.com/SciTools/iris/pull/5956#discussion_r1654598764

Reply:
I've decided against this.

The principle laid out by @bouweandela - don't rely on private objects in third party libraries - is definitely valid, and I'm aware I'm taking a risk here. However the alternative of carefully controlling the arguments passed to the ASV class relies on knowledge of ASV internals; I have that knowledge, but there is no contract that it will be stable between patch releases (the same problem as private objects like `_path`).

We have coped with breaking changes in ASV patch releases before, and I'm prepared to keep that risk since there is no clear, safer alternative.

Thankfully this new implementation does not fail MyPy around `self._path` anyway.